envoy/cilium/npds.proto

```diff
@@ -34,62 +34,67 @@ message NetworkPolicy {
+  // Every PortNetworkPolicy element in this set has a unique port / protocol
+  // combination.
+  // Optional. If empty, all flows in this direction are denied.
+  repeated PortNetworkPolicy egress_per_port_policies = 3;
 }
 
-// A network policy to be enforced on flows to a specific destination L4 port.
+// A network policy to whitelist flows to a specific destination L4 port,
+// as a conjunction of predicates on L3/L4/L7 flows.
+// If all the predicates of a policy match a flow, the flow is whitelisted.
 message PortNetworkPolicy {
   // The flows' destination L4 port number, as an unsigned 16-bit integer.
-  // Required.
+  // Optional. If not specified, all destination L4 port numbers are matched by
+  // this predicate.
   uint32 port = 1 [(validate.rules).uint32.lte = 65535];
 
   // The flows' L4 transport protocol.
-  // Required.
+  // Optional. If not specified, all L4 transport protocols are matched by this
+  // predicate.
   envoy.api.v2.core.SocketAddress.Protocol protocol = 2;
-  // TODO: Fix protoc-gen-validate. Adding any validate.rules to protocol above results in generating broken C++ code.
 
   // The network policy rules to be enforced on the flows to the port.
-  // A flow is allowed if any rule matches it. If no rule matches the flow, it is denied.
-  // Required and may not be empty.
-  repeated PortNetworkPolicyRule rules = 3 [(validate.rules).repeated .min_items = 1];
+  // Optional. A flow is matched by this predicate if either the set of
+  // rules is empty or any of the rules matches it.
+  repeated PortNetworkPolicyRule rules = 3;
 }
 
-// A network policy rule, as a conjunction of predicates on L3-L7 flows.
-// If all the predicates of a rule match a flow, the flow is allowed.
+// A network policy rule, as a conjunction of predicates on L3/L7 flows.
+// If all the predicates of a rule match a flow, the flow is matched by the
+// rule.
 message PortNetworkPolicyRule {
-  // The set of identifiers of policies of remote hosts that are allowed by this rule.
-  // This predicate matches a flow if the identifier of the policy applied on the flow's remote
-  // host is contained in this set. Required and may not be empty.
-  repeated uint64 remote_policies = 1 [(validate.rules).repeated = {min_items: 1, unique: true}];
+  // The set of identifiers of policies of remote hosts.
+  // A flow is matched by this predicate if the identifier of the policy
+  // applied on the flow's remote host is contained in this set.
+  // Optional. If not specified, any remote host is matched by this predicate.
+  repeated uint64 remote_policies = 1 [(validate.rules).repeated.unique = true];
 
   oneof l7_rules {
     // The set of HTTP network policy rules.
-    // An HTTP request is allowed if any rule in this set matches it.
-    // Optional.
+    // An HTTP request is matched by this predicate if any of its rules matches
+    // the request.
+    // Optional. If not specified, any HTTP request is matched by this
+    // predicate.
```

Comment:
As on one-of, we may not be able to deduce "HTTP" from a missing field. To match any HTTP we likely need to include this, but have the contents be empty.

Reply:
Moved the "Optional" comment to the oneof itself.